Incorrect error codes reported

**Version Used**: 

I'm updating Windows Forms SDK/toolset from P6 to RC1, and our builds fail with errors which we have already suppressed. E.g.:
![image](https://user-images.githubusercontent.com/4403806/93298989-a39bae00-f837-11ea-8a92-72d19aec649d.png)

It appears instead of `CS0618` we must now be suppressing `SYSLIB0011` but I had to spend time to figure it out on my own.
Likewise we have other failures, e.g.
![image](https://user-images.githubusercontent.com/4403806/93302561-89fd6500-f83d-11ea-99c5-c2a65a31ecbb.png)
...and only VS is telling me that I now need to suppress a different one:
![image](https://user-images.githubusercontent.com/4403806/93302704-c0d37b00-f83d-11ea-9db1-1b8bfaef18ea.png)

Few builds:
* https://dev.azure.com/dnceng/public/_build/results?buildId=817132&view=results
* https://dev.azure.com/dnceng/public/_build/results?buildId=817156&view=results

**Steps to Reproduce**:

1. Clone this commit: https://github.com/dotnet/winforms/commit/2427ee0140db52050535e9459384e6527610720d
2. Run `.\build.cmd`
3. Observe multiple failures

**Expected Behavior**:

Errors must either correctly report codes or use the existing suppression rules.

From cli I get this:
```
 error CS1029: #error: 'version' [C:\Development\winforms.temp\src\System.Windows.Forms.Design\src\System.Windows.Forms.Design.csproj]
 error CS8304: Compiler version: '3.8.0-3.20452.3 (90e570d7)'. Language version: preview. 
```

`/+warnaserror` seems to be the problem. The minimal repro project https://github.com/dotnet/roslyn/issues/47736#issuecomment-693708428 gives an unexpected `error CS0618` when either `<WarningsAsErrors>SYSLIB0011</WarningsAsErrors>` is specified or `<TreatWarningsAsErrors>true</TreatWarningsAsErrors>` is specified.

Thanks to @jmarolf for helping me track this down. I suppose this bug didn't have the chance to negatively impact more people since corefx didn't add custom diagnostic IDs until recently.

--- a/src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs
+++ b/src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs
@@ -4,6 +4,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Linq;
 using System.Reflection;
 using System.Reflection.Metadata;
@@ -8682,35 +8683,78 @@ class C2 : C1
             var comp1 = CreateCompilation(new[] { ObsoleteAttributeSource, source1 });
             comp1.VerifyDiagnostics();
 
-            verify(comp1.ToMetadataReference(), TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Warn),
-                // (6,9): warning TEST1: 'C1.M1()' is obsolete
-                //         M1(); // 1
-                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9)
-                );
-            verify(comp1.EmitToImageReference(), TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Warn),
+            // WithGeneralDiagnosticOption
+            verify(TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Warn),
                 // (6,9): warning TEST1: 'C1.M1()' is obsolete
                 //         M1(); // 1
                 Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9)
                 );
 
-            verify(comp1.ToMetadataReference(), TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Error),
-                // (6,9): error TEST1: 'C1.M1()' is obsolete
-                //         M1(); // 1
-                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9).WithWarningAsError(true)
-                );
-            verify(comp1.EmitToImageReference(), TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Error),
+            verify(TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Error),
                 // (6,9): error TEST1: 'C1.M1()' is obsolete
                 //         M1(); // 1
                 Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9).WithWarningAsError(true)
                 );
 
-            void verify(MetadataReference reference, CSharpCompilationOptions options, params DiagnosticDescription[] expectedDiagnostics)
+            verify(TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Hidden),
+                // (6,9): warning TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9));
+
+            verify(TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Suppress));
+
+
+            // WithSpecificDiagnosticOption for id TEST1
+            verify(TestOptions.DebugDll.WithSpecificDiagnosticOptions(ImmutableDictionary<string, ReportDiagnostic>.Empty.Add("TEST1", ReportDiagnostic.Warn)),
+                // (6,9): warning TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9)
+                );
+
+            verify(TestOptions.DebugDll.WithSpecificDiagnosticOptions(ImmutableDictionary<string, ReportDiagnostic>.Empty.Add("TEST1", ReportDiagnostic.Error)),
+                // (6,9): error TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9).WithWarningAsError(true)
+                );
+
+            verify(TestOptions.DebugDll.WithSpecificDiagnosticOptions(ImmutableDictionary<string, ReportDiagnostic>.Empty.Add("TEST1", ReportDiagnostic.Hidden)),
+                // (6,9): hidden TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9)
+                );
+
+            verify(TestOptions.DebugDll.WithSpecificDiagnosticOptions(ImmutableDictionary<string, ReportDiagnostic>.Empty.Add("TEST1", ReportDiagnostic.Suppress)));
+
+
+            // WithSpecificDiagnosticOption for id CS0618
+            verify(TestOptions.DebugDll.WithSpecificDiagnosticOptions(ImmutableDictionary<string, ReportDiagnostic>.Empty.Add("CS0618", ReportDiagnostic.Error)),
+                // (6,9): warning TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9)
+                );
+
+            verify(TestOptions.DebugDll.WithSpecificDiagnosticOptions(ImmutableDictionary<string, ReportDiagnostic>.Empty.Add("CS0618", ReportDiagnostic.Suppress)),
+                // (6,9): warning TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9)
+                );
+
+            void verify(CSharpCompilationOptions options, params DiagnosticDescription[] expectedDiagnostics)
+            {
+                verifyReference(comp1.ToMetadataReference(), options, expectedDiagnostics);
+                verifyReference(comp1.EmitToImageReference(), options, expectedDiagnostics);
+            }
+
+            void verifyReference(MetadataReference reference, CSharpCompilationOptions options, DiagnosticDescription[] expectedDiagnostics)
             {
                 var comp2 = CreateCompilation(source2, references: new[] { reference }, options: options);
                 var diags = comp2.GetDiagnostics();
 
-                var diag = diags.Single();
-                Assert.Equal("https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/TEST1", diag.Descriptor.HelpLinkUri);
+                if (expectedDiagnostics.Any())
+                {
+                    var diag = diags.Single();
+                    Assert.Equal("https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/TEST1", diag.Descriptor.HelpLinkUri);
+                }
 
                 diags.Verify(expectedDiagnostics);
             }

--- a/src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs
+++ b/src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs
@@ -8658,6 +8658,65 @@ class C2 : C1
         }
 
         [Fact, WorkItem(42119, "https://github.com/dotnet/roslyn/issues/42119")]
+        public void Obsolete_CustomDiagnosticId_FromMetadata_05()
+        {
+            var source1 = @"
+using System;
+#pragma warning disable 436
+
+public class C1
+{
+    [Obsolete(DiagnosticId = ""TEST1"", UrlFormat = ""https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/{0}"")]
+    public void M1() { }
+}
+";
+
+            var source2 = @"
+class C2 : C1
+{
+    void M2()
+    {
+        M1(); // 1
+    }
+}";
+            var comp1 = CreateCompilation(new[] { ObsoleteAttributeSource, source1 });
+            comp1.VerifyDiagnostics();
+
+            verify(comp1.ToMetadataReference(), TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Warn),
+                // (6,9): warning TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9)
+                );
+            verify(comp1.EmitToImageReference(), TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Warn),
+                // (6,9): warning TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9)
+                );
+
+            verify(comp1.ToMetadataReference(), TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Error),
+                // (6,9): error TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9).WithWarningAsError(true)
+                );
+            verify(comp1.EmitToImageReference(), TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Error),
+                // (6,9): error TEST1: 'C1.M1()' is obsolete
+                //         M1(); // 1
+                Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9).WithWarningAsError(true)
+                );
+
+            void verify(MetadataReference reference, CSharpCompilationOptions options, params DiagnosticDescription[] expectedDiagnostics)
+            {
+                var comp2 = CreateCompilation(source2, references: new[] { reference }, options: options);
+                var diags = comp2.GetDiagnostics();
+
+                var diag = diags.Single();
+                Assert.Equal("https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/TEST1", diag.Descriptor.HelpLinkUri);
+
+                diags.Verify(expectedDiagnostics);
+            }
+        }
+
+        [Fact, WorkItem(42119, "https://github.com/dotnet/roslyn/issues/42119")]
         public void Obsolete_CustomDiagnosticId_BadMetadata_01()
         {
             var source1 = @"

--- a/src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs
+++ b/src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_WellKnownAttributes.cs
@@ -8697,7 +8697,7 @@ class C2 : C1
                 );
 
             verify(TestOptions.DebugDll.WithGeneralDiagnosticOption(ReportDiagnostic.Hidden),
-                // (6,9): warning TEST1: 'C1.M1()' is obsolete
+                // (6,9): hidden TEST1: 'C1.M1()' is obsolete
                 //         M1(); // 1
                 Diagnostic("TEST1", "M1()", isSuppressed: false).WithArguments("C1.M1()").WithLocation(6, 9));
 

--- a/src/Compilers/Core/Portable/Diagnostic/CustomObsoleteDiagnosticInfo.cs
+++ b/src/Compilers/Core/Portable/Diagnostic/CustomObsoleteDiagnosticInfo.cs
@@ -23,6 +23,12 @@ namespace Microsoft.CodeAnalysis
             Data = data;
         }
 
+        private CustomObsoleteDiagnosticInfo(CustomObsoleteDiagnosticInfo baseInfo, DiagnosticSeverity effectiveSeverity)
+            : base(baseInfo, effectiveSeverity)
+        {
+            Data = baseInfo.Data;
+        }
+
         public override string MessageIdentifier
         {
             get
@@ -48,6 +54,11 @@ namespace Microsoft.CodeAnalysis
 
                 return _descriptor;
             }
+        }
+
+        internal override DiagnosticInfo GetInstanceWithSeverity(DiagnosticSeverity severity)
+        {
+            return new CustomObsoleteDiagnosticInfo(this, severity);
         }
 
         private DiagnosticDescriptor CreateDescriptor()

--- a/src/Compilers/Core/Portable/Diagnostic/DiagnosticInfo.cs
+++ b/src/Compilers/Core/Portable/Diagnostic/DiagnosticInfo.cs
@@ -62,7 +62,7 @@ namespace Microsoft.CodeAnalysis
             _arguments = arguments;
         }
 
-        private DiagnosticInfo(DiagnosticInfo original, DiagnosticSeverity overriddenSeverity)
+        protected DiagnosticInfo(DiagnosticInfo original, DiagnosticSeverity overriddenSeverity)
         {
             _messageProvider = original.MessageProvider;
             _errorCode = original._errorCode;
@@ -137,7 +137,7 @@ namespace Microsoft.CodeAnalysis
         }
 
         // Create a copy of this instance with a explicit overridden severity
-        internal DiagnosticInfo GetInstanceWithSeverity(DiagnosticSeverity severity)
+        internal virtual DiagnosticInfo GetInstanceWithSeverity(DiagnosticSeverity severity)
         {
             return new DiagnosticInfo(this, severity);
         }

